grails-app/controllers/org/pih/warehouse/inventory/StockMovementController.groovy

```diff
@@ -267,7 +268,30 @@ class StockMovementController {
     }
 
     def rollbackApproval = {
-        throw new NotImplementedException("Action ${params.actionName} not implemented")
+        String stockMovementId = params.get("id")
+        if (!stockMovementService.canRollbackApproval(AuthService.currentUser.get(), stockMovementId)) {
+            flash.message = g.message(
+                    code: "request.errors.approval.rollback.message",
+                    default: "Unable to rollback approval: You do not have permissions to perform this action",
+                    args: [g.message(code: "errors.noPermissions.label")]
+            )
+        }
+
+        try {
+            stockMovementService.rollbackApproval(stockMovementId)
+            flash.message = g.message(
+                    code: "request.successfulApproval.rollback.message",
+                    default: "Successfully rolled back approval",
+            )
+        } catch (Exception e) {
```

Comment:
I also feel hesitant about this `catch`, because it might prevent potential rollback from working (I don't remember if it was Grails 1 or Grails 3 that was not rolling back when checked exception happens).
Since you do a few operations, like - deleting an event, updating a requisition (setting some properties to `null`), I think something weird might happen, but I would have to debug it myself to confirm that. Just wanted to point that out.

Reply:
@jmiranda I know it is. I meant that weird things might happen if the **transaction rollback** would happen.